infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/RoutingQueuesSnapshotReader.java

```diff
@@ -244,22 +251,30 @@ public void requestClientForSnapshotData(UUID snapshotRequestId) {
             log.error("Unrecoverable exception while adding enforce snapshot sync event", e);
             throw new UnrecoverableCorfuInterruptedError(e);
         }
+        lastDataReceivedEpoch = System.currentTimeMillis();
     }
 
-    // Waits for a default timeout period for data to get written for this destination on the Snapshot Sync queue
-    private void waitForData() {
-        // Create a task to wait for more data until DATA_WAIT_TIMEOUT_MS
-        StreamQueryTask streamQueryTask = new StreamQueryTask();
-        Future<Void> queryFuture = dataPoller.submit(streamQueryTask);
-
+    private boolean dataFound() {
         try {
-            queryFuture.get(DATA_WAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        } catch (TimeoutException e) {
-            throw new ReplicationReaderException("Timed out waiting for data or end marker for Snapshot Sync", e);
+            buildOpaqueStreamIterator();
+            if (currentStreamHasNext()) {
+                return true;
+            }
         } catch (Exception e) {
-            // Handle all other types of exceptions
-            log.error("Caught exception in WaitForData ", e);
+            log.error("Unexpected exception caught. ", e);
         }
+
+        log.info("Retrying. currentStreamInfo={}", currentStreamInfo.maxVersion);
+        return false;
+    }
+
+    @VisibleForTesting
+    public boolean receiveWindowTimedOut() {
+        // This for unit test. In real scenarios, lastDataReceivedEpoch would not be 0 when we reach here.
```

Comment:
It can be 0 during the very first run, right?

Reply:
No, we assign it the current time when we request for snapshot data from the client. So the clock starts from the time we request for data...and incase we do not receive any data within the timeout window, we can place a request again.